File: src/java/magma/result/Ok.java
package magma.result;

import java.util.function.Function;

public record Ok<T, X>(T value) implements Result<T, X> {
    @Override
    public <R> Result<R, X> flatMapValue(Function<T, Result<R, X>> mapper) {
        return mapper.apply(this.value);
    }

    @Override
    public <R> Result<R, X> mapValue(Function<T, R> mapper) {
        return new Ok<>(mapper.apply(this.value));
    }

    @Override
    public <R> Result<T, R> mapErr(Function<X, R> mapper) {
        return new Ok<>(this.value);
    }

    @Override
    public <R> R match(Function<T, R> onOk, Function<X, R> onErr) {
        return onOk.apply(this.value);
    }
}

File: src/c/magma/result/Ok.c
#include <temp.h>
#include <temp.h>
#include <temp.h>
struct Ok<T, X> {
	T value;
	struct Ok<T, X> Ok<T, X>_new(T value){
		struct Ok<T, X> this;
		this.value = value;
		return this;
	}
	X> Ok<T, X>_flatMapValue(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		return mapper.apply(this.value);
	}
	X> Ok<T, X>_mapValue(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		temp = temp;
	}
	R> Ok<T, X>_mapErr(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		temp = temp;
	}
	R Ok<T, X>_match(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		return onOk.apply(this.value);
	}
	X> Ok<T, X>_and(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		return other.get().mapValue(otherValue -> new Tuple<>(this.value, otherValue));
	}
	R>> Ok<T, X>_or(void* _this_){
		struct Ok<T, X> this = *(struct Ok<T, X>*) this;
		temp = temp;
	}
};